### DISCUSSION TITLE
How to add custom heads?

### DISCUSSION THREAD

**Discussion by g0urra** [NillerMedDild/Enigmatica6#3786] [Q&A]
So I'm looking into adding more heads that are obtained by the Severing modifier from Tinker's Construct, which lets you decapitate the enemy and obtain their skull when killed. I've looked into how find these types and found these in the .json file:

`\data\tconstruct\loot_tables\blocks\<x>_head.json` (something related to the head block)
`\data\forge\tags\items\heads.json` (lists the heads added by the mod)
`\data\tconstruct\recipes\tools\severing\<x>_head.json` (probably adds that the mob in question can drop the head when slain with Severing)
`\assets\tconstruct\blockstates\<x>_head.json` (makes the base block a skull that the head textures are put on?)
`\assets\tconstruct\models\item\<x>_head.json` (something related to the above, but with the asset)
`\assets\tconstruct\textures\entity\skull` (folder than contains the skull/head textures)
`\data\tconstruct\recipes\smeltery\entity_melting\heads\<x>.json` (adds results when the head is melted in a Smeltery/Foundry)

I could look into making these files, but I have no idea how to add them to the modpack.

**Comment by TemperedFool** [comment]
Is this for heads that exist and just don't have severing applied, or would we need to add the head blocks/items too?

I'm going to guess that we can achieve this with the KubeJS loot pack injection we've added (see https://github.com/NillerMedDild/Enigmatica6/blob/01c06699f9ad0edb0aa14e9a7a04a321526e827d/kubejs/server_scripts/enigmatica/kubejs/base/loot_tables/atum/chests/pharaoh.js for an example).

I'll peek at those files a bit, and see if it's anymore complicated than that. Someone else feel free to beat me to it ;-)

**Reply by g0urra** [reply]
It's for heads that don't exist yet, but at the top of my head these are some ideas: Brutish Zombie, Wraith, Mummy, Pharaoh.

**Reply by TemperedFool** [reply]
Hmm, ok. In that case, what would the heads be used for? We'd also need art assets for them. Creating the items themselves isn't too tough, would be done via the block and item registries here: https://github.com/NillerMedDild/Enigmatica6/tree/master/kubejs/startup_scripts

Looking through the repo for TC, it looks like there may be a beheading recipe that we can datapack. This looks (I think) like Craft Tweaker scripts: https://github.com/SlimeKnights/TinkersConstruct/blob/1.16/src/main/resources/data/tconstruct/scripts/tconstruct/severing.zs

If that's the case, I can figure out what the recipe type is, adding a script for it should be simple.

**Reply by Darkere** [reply]
Figure out what we want and talk to slimeknight about this and save yourself hours of work.

**Reply by TemperedFool** [reply]
Why do that when I can be stubborn about doing it myself and dig through their repo😉

Here is one of the generated recipes; should be super simple, if we have the items:

https://github.com/SlimeKnights/TinkersConstruct/blob/3ecadbec521852865ad5b3fff7f2649862941adf/src/generated/resources/data/tconstruct/recipes/tools/severing/blaze_head.json

**Reply by g0urra** [reply]
> Hmm, ok. In that case, what would the heads be used for? We'd also need art assets for them.
> Creating the items themselves isn't too tough, would be done via the block and item registries here: https://github.com/NillerMedDild/Enigmatica6/tree/master/kubejs/startup_scripts

Creating the art assets wouldn't be too difficult I would imagine, as they can be based on the textures of the mobs in question. For the scripts it would be a matter of making new ones based on the code and naming of current ones.

> Looking through the repo for TC, it looks like there may be a beheading recipe that we can datapack. This looks (I think) like Craft Tweaker scripts: https://github.com/SlimeKnights/TinkersConstruct/blob/1.16/src/main/resources/data/tconstruct/scripts/tconstruct/severing.zs
> If that's the case, I can figure out what the recipe type is, adding a script for it should be simple.

I'm not sure if it's on Github (I could not find it) but it's in the mod's .json file.

**Reply by TemperedFool** [reply]
> I'm not sure if it's on Github (I could not find it) but it's in the mod's .json file.

I linked to the recipe json in the comment above yours 😉

**Reply by TemperedFool** [reply]
The recipe script would go in the same place as this one, and be structured similarly (but adapted for the recipe JSON structure): https://github.com/NillerMedDild/Enigmatica6/blob/76549e9fa7aaf92977bc837656024589b96669b8/kubejs/server_scripts/enigmatica/kubejs/base/recipetypes/tconstruct/melting_fuel.js

**Reply by g0urra** [reply]
> Here is one of the generated recipes; should be super simple, if we have the items:
> 
> https://github.com/SlimeKnights/TinkersConstruct/blob/3ecadbec521852865ad5b3fff7f2649862941adf/src/generated/resources/data/tconstruct/recipes/tools/severing/blaze_head.json

Doh, I should have dug a bit deeper :P In the case of Blaze Head, these are some of the links (in order):

- https://github.com/SlimeKnights/TinkersConstruct/blob/3ecadbec521852865ad5b3fff7f2649862941adf/src/generated/resources/data/tconstruct/loot_tables/blocks/blaze_head.json
- https://github.com/SlimeKnights/TinkersConstruct/blob/3ecadbec521852865ad5b3fff7f2649862941adf/src/generated/resources/data/tconstruct/recipes/tools/severing/blaze_head.json
- https://github.com/SlimeKnights/TinkersConstruct/blob/3ecadbec521852865ad5b3fff7f2649862941adf/src/generated/resources/data/tconstruct/recipes/smeltery/entity_melting/heads/blaze.json

The assets directories however seem to be inaccessible through Github?

**Reply by TemperedFool** [reply]
Here maybe: https://github.com/SlimeKnights/TinkersConstruct/tree/1.16/src/main/resources/assets/tconstruct/textures/entity/skull

**Reply by TemperedFool** [reply]
The rest may be generated in code, not sure

**Reply by TemperedFool** [reply]
@MaxNeedsSnacks - how complicated would making a custom skull (assuming we have textures) be with Kube, if it's possible at all?

**Reply by g0urra** [reply]
Did we get an answer about this?
